README.md
evilscan
===============

![Node.js CI](https://github.com/eviltik/evilscan/workflows/Node.js%20CI/badge.svg)
[![npm version](https://badge.fury.io/js/evilscan.svg?1.8.0)](https://badge.fury.io/js/evilscan)
[![MIT Licence](https://badges.frapsoft.com/os/mit/mit.svg?v=103)](https://opensource.org/licenses/mit-license.php)
[![Depfu](https://badges.depfu.com/badges/60413e5ea875bbdff97a0b099a948e75/overview.svg)](https://depfu.com/github/eviltik/evilscan?project_id=22885)

Nodejs Simple Network Scanner


Features
--------
* individual IP or IP range scan
* individual port, ports list, or ports range
* banner grabbing (not fully implemented, works with native verbose ports only)
* IAC negotiation for telnet
* reverse dns
* geolocation information
* stdout or json output
* optional progress details (event based)


Install
-------

>```
>npm install -g evilscan
>```


Usage: in your code, using events only
-------------------
```
const Evilscan = require('evilscan');

const options = {
    target:'127.0.0.1',
    port:'21-23',
    status:'TROU', // Timeout, Refused, Open, Unreachable
    banner:true
};

const evilscan = new Evilscan(options);

evilscan.on('result',data => {
    // fired when item is matching options
    console.log(data);
});

evilscan.on('error', err => {
    throw new Error(data.toString());
});

evilscan.on('done', () => {
    // finished !
});

evilscan.run();
```


Usage: in your code, using callback and events
-------------------
```
const Evilscan = require('evilscan');

const options = {
    target:'127.0.0.1',
    port:'21-23',
    status:'TROU', // Timeout, Refused, Open, Unreachable
    banner:true
};

new Evilscan(options, (err, scan) => {

    if (err) {
        console.log(err);
        return;
    }

    scan.on('result', data => {
        // fired when item is matching options
        console.log(data);
    });

    scan.on('error', err => {
        throw new Error(data.toString());
    });

    scan.on('done', () => {
        // finished !
    });

    scan.run();
});

```

Usage: command line
-------------------
Usage: evilscan <fqdn|ipv4|cidr> [options]

Example:
```
evilscan 192.168.0.0/24 --port=21-23,80
```


Options
-------
```
  --port          port(s) you want to scan, examples:
                  --port=80
                  --port=21,22
                  --port=21,22,23,5900-5902

  --reverse       display DNS reverse lookup

  --reversevalid  only display results having a valid reverse dns, except if
                  ports specified

  --geo           display geoip (free maxmind)

  --banner        display banner

  --bannerlen     set banner length grabing
                  default 512

  --bannerraw     display raw banner (as a JSON Buffer)

  --progress      display progress indicator each seconds

  --status        ports status wanted in results (example --status=OT)
                  T(timeout)
                  R(refused)
                  O(open, default)
                  U(unreachable)

  --scan          scan method
                  tcpconnect (full connect, default)
                  tcpsyn (half opened, not yet implemented)
                  udp (not yet implemented)

  --concurrency   max number of simultaneous socket opened
                  default 500

  --timeout       maximum number of milliseconds before closing the connection
                  default 2000

  --display       display result format (json,xml,console)
                  default console

  --infile        fetch fqdn/ipv4/ipv6 target from a file
  
  --outfile       dump result in a file

  --json          shortcut for --display=json

  --xml           shortcut for --display=xml

  --console       shortcut for --display=console

  --help          display help

  --about         display about

  --version       display version number
```

Samples output
----------------

* Every ports on localhost, grab banner, display only opened ports
```
eviltik@debian:~# evilscan 127.0.0.1 --port=0-65535 --banner
127.0.0.1|111||open
127.0.0.1|53||open
127.0.0.1|23|Debian GNU/Linux jessie/sid\r\ndebian login:|open
127.0.0.1|5432||open
127.0.0.1|27017||open
127.0.0.1|28017||open
127.0.0.1|35223||open
127.0.0.1|35491||open
127.0.0.1|39619||open
done, 9 result(s)
```

* Every ports on localhost, grab banner, display only opened ports, json output, progress status each seconds
```
eviltik@debian:~# evilscan 127.0.0.1 --port=0-65535 --banner --isopen --istimeout --progress --json
{"_timeStart":"N/A","_timeElapsed":"N/A","_jobsTotal":65535,"_jobsRunning":0,"_jobsDone":0,"_progress":0,"_concurrency":500,"_status":"Starting","_message":"Starting"}
{"_timeStart":1371245901876,"_timeElapsed":1031,"_jobsTotal":65535,"_jobsRunning":500,"_jobsDone":7638,"_progress":11,"_concurrency":500,"_status":"Running","_message":"Scanning 127.0.0.1:8138"}
{"_timeStart":1371245901876,"_timeElapsed":2085,"_jobsTotal":65535,"_jobsRunning":500,"_jobsDone":16137,"_progress":24,"_concurrency":500,"_status":"Running","_message":"Scanning 127.0.0.1:16637"}
{"ip":"127.0.0.1","port":111,"status":"open"}
{"ip":"127.0.0.1","port":53,"status":"open"}
{"ip":"127.0.0.1","port":23,"banner":"Debian GNU/Linux jessie/sid\\r\\ndebian login:","status":"open"}
{"ip":"127.0.0.1","port":5432,"status":"open"}
{"_timeStart":1371245901876,"_timeElapsed":3107,"_jobsTotal":65535,"_jobsRunning":500,"_jobsDone":24656,"_progress":37,"_concurrency":500,"_status":"Running","_message":"Scanning 127.0.0.1:25156"}
{"_timeStart":1371245901876,"_timeElapsed":4166,"_jobsTotal":65535,"_jobsRunning":500,"_jobsDone":33166,"_progress":50,"_concurrency":500,"_status":"Running","_message":"Scanning 127.0.0.1:33666"}
{"_timeStart":1371245901876,"_timeElapsed":5215,"_jobsTotal":65535,"_jobsRunning":500,"_jobsDone":41664,"_progress":63,"_concurrency":500,"_status":"Running","_message":"Scanning 127.0.0.1:42164"}
{"ip":"127.0.0.1","port":27017,"status":"open"}
{"ip":"127.0.0.1","port":28017,"status":"open"}
{"_timeStart":1371245901876,"_timeElapsed":6217,"_jobsTotal":65535,"_jobsRunning":500,"_jobsDone":49682,"_progress":75,"_concurrency":500,"_status":"Running","_message":"Scanning 127.0.0.1:50182"}
{"ip":"127.0.0.1","port":35491,"status":"open"}
{"ip":"127.0.0.1","port":35223,"status":"open"}
{"ip":"127.0.0.1","port":39619,"status":"open"}
{"_timeStart":1371245901876,"_timeElapsed":7234,"_jobsTotal":65535,"_jobsRunning":500,"_jobsDone":57732,"_progress":88,"_concurrency":500,"_status":"Running","_message":"Scanning 127.0.0.1:58232"}
{"_timeStart":1371245901876,"_timeElapsed":8182,"_jobsTotal":65535,"_jobsRunning":0,"_jobsDone":65535,"_progress":100,"_concurrency":500,"_status":"Finished","_message":"Scanning 127.0.0.1:65535"}
```

Tips :
--------
**Concurrency and fast scan**

By default, concurrency is 100. Which is slow when you are scanning large ip range or large port range. You can pass a bigger value using --concurrency option. 1000 is fast by example.
On some linux, only 1024 opened sockets are allowed in the same time. To break this limit, you have to update ulimit parameter of your linux first :

```
ulimit -u unlimited
```

In all cases, due to #25, you will not be able to scan more than 16580355 ipv4 addresses at the moment.


**Pause/unpause**

You can pause/unpause a running scan by sending SIGUSR2 signal. First time it will pause the process, second time it will unpause it.
```
kill -SIGUSR2 19859 # where 19859 is the pid of nodejs process running evilscan
```


changelog.md
v1.9.1 - 2022/04/04
* fix README example
  
v1.9.0 - 2022/04/04
* add #32 (single ipv6 support)
* maintenance (upgrade node packages)

v1.8.1 - 2021/08/30
* maintenance (upgrade node packages)
  
v1.8.0 - 2021/03/01
* refactor
* close #59
 
v1.7.8 - 2021/01/25
* remove travis-ci, add github ci

v1.7.6 - 2021/01/24
* fix #61 (badge GPL icon in the readme)
* fix #60 (add infile and outfile option)

v1.7.5 - 2019/01/23
* switch to MIT licence

v1.7.4 - 2019/01/23
* fix #57

v1.7.3 - 2019/01/23
* maintenance: update npm packages
* fix geoip test
* ci tests should accept nodejs v10 and v11

v1.7.2 - 2019/01/23 (broken tests, see 1.7.3)
* don't destroy a non existing socket, fix #58

v1.7.1 - 2018/09/29
* don't require geoip-lite if geo option not specified (default) fix #54

v1.1.0 => v1.7.0
* i was probably drunked, i don't remember

v1.1.0 - 2017/11/12
* update some package dependencies (warn, see #48)

v1.0.3 - 2016/11/26
* fix #43

v1.0.2 - 2016/03/13
* fix #39 (display ip list when no port specified)
* merge #38 (async version)
* fix SPDX license expression in package.json
* use static version module in package.json

v1.0.1 - 2015/03/03

v1.0.0 - forgot to update changelog, omfg
* release 1.0

v0.0.9 - forgot to update changelog, shit

v0.0.8 - 2013/06/16
* fix #23 (SIGUSR2 rather than SIGUSR1)
* fix #24 (done event fired twice)
* fix #7 (pause/unpause test)

v0.0.7 - 2013/06/15
* (no ticket) fix --help

v0.0.6 - 2013/06/15
* #14 cache DNS
* #14 cache Geoip
* #1 banner length option

v0.0.5 - 2013/06/14
===================
* #12 complete refactor
* Breaking changes regarding command line options, sorry for that
* #12 can now be used as a node module, event based
* #11 add --help options
* Add changelog.md
* Add tests

v0.0.4 - 2013/06/11
===================
* First npm module
* Really bad code


index.js
const EventEmitter = require('events').EventEmitter;
const TcpScan = require('./libs/TcpScan');
const options = require('./libs/options');
const dns = require('dns');
const async = require('async');

var geoip;

class Evilscan extends EventEmitter {

    constructor(opts, callback) {
        super();

        this.lastMessage = 'Starting';
        this.paused = false;
        this.progress = 0;
        this.progressTimer = null;
        this.info = {
            nbIpToScan:0,
            nbPortToScan:0
        };
        this.cacheGeo = {};
        this.cacheDns = {};

        this.options = opts || {};
        this.q = null;

        if (!opts.ips) {
            // Called from a js script, reparse options
            options.parse(opts, (err, opts) => {

                if (err) {
                    // @TODO: is there a way to know if error handler has been
                    // defined rather than this shit of try catch ?
                    try {
                        return this.emit('error', err);
                    } catch(e) {
                        //
                    }
                }

                if (err && callback) {
                    return callback(err, this);
                }

                this.options = opts;
                this._init();
                callback && callback(null, this);
            });
            return;
        }

        this._init();
    }

    _init() {

        //if (this.options && this.options.concurrency) {
        this.q = require('qjobs')({
            maxConcurrency:this.options.concurrency||500
        });

        this._initTestLocalhost();
        this._initQueue();
        this._initQueueProgress();
        this._initQueuePause();
        //portscan.setSocketTimeout(argv.timeout||1000);
    }

    _initTestLocalhost() {
        // For tests, the telnet server can not
        // accept more than 50 simultaneous connection,
        // so let's make a pause between each pools

        if (!this.options || !this.options.ips || !this.options.ips.length) {
            return;
        }

        if (
            this.options.ips.length == 1
            &&
            this.options.ips[0].match(/^127.0.0.1\//)
        ) {
            this.q.setInterval(1000);
        }
    }

    _initQueue() {
        // Push scan jobs in the queue

        if (!this.options || !this.options.ips || !this.options.ips.length) {
            return;
        }

        const maxi = this.options.ips.length;
        let maxj = this.options.ports.length;
        const maxt = maxi*maxj;
        let i = 0;
        let j = 0;
        let ip;
        const copyOfIps = JSON.parse(JSON.stringify(this.options.ips));

        this.info.combinaison = maxt;

        while (copyOfIps.length) {
            ip = copyOfIps.shift();
            if (this.options.ports.length) {
                maxj = this.options.ports.length;
                for (j=0;j<maxj;j++) {
                    const args = {
                        ip,
                        port:this.options.ports[j],
                        i,
                        max:maxt
                    };
                    this.info.nbPortToScan++;
                    this.q.add(this.scan.bind(this), args);
                    i++;
                }
                this.info.nbIpToScan++;
            } else {
                const args = {
                    ip,
                    i,
                    max:maxt
                };
                this.q.add(this.scan.bind(this), args);
                this.info.nbPortToScan = 0;
                this.info.nbIpToScan = 1;
                i++;
            }
        }

        this.q.on('end', () => {
            this.cacheGeo = {};
            this.cacheDns = {};
            this.emit('done');
            this.callbackRun && this.callbackRun();
        });

        this.q.on('jobStart', data => {
            let str = 'Scanning '+data.ip;
            if (data.port) str+=':'+data.port;
            this.lastMessage = str;
            this.emit('scan', data);
        });

        this.q.on('jobEnd', data => {
            let str = 'Scanned '+data.ip;
            if (data.port) str+=':'+data.port;
            this.lastMessage = str;
        });
    }

    _initQueueProgress() {
        if (!this.options.progress) {
            return;
        }

        this.progressTimer = setInterval(this._triggerQueueProgress.bind(this), 1000);

        this.q.on('start', () => {
            this.emit('start');
            this._triggerQueueProgress();
        });

        this.q.on('end', () => {
            clearInterval(this.progressTimer);
            this._triggerQueueProgress();
            this.callback && this.callback();
        });
    }

    _initQueuePause() {

        if (process.eventNames().indexOf('SIGUSR2') >= 0) {
            // avoid MaxListenersExceededWarning
            // see https://github.com/eviltik/evilscan/issues/41#issuecomment-364630079
            // thanks John cetfor for pointing this ;)
            return;
        }

        process.on('SIGUSR2', () => {
            if (!this.paused) {
                this.paused = true;
                this.lastMessage = 'Pause';
            } else {
                this.paused = false;
            }
            this.q.pause(this.paused);
        });
    }

    _triggerQueueProgress() {
        var o = this.q.stats();
        if (!this.paused) {
            o._message = this.lastMessage;
        } else {
            o._message = 'Paused';
            o._status = 'Paused';
        }

        this.progress = o._progress;
        this.emit('progress', o, this);
    }

    _geolocate(ip, callback) {

        if (!this.options.geo) {
            return callback();
        }

        if (!geoip) {
            geoip = require('geoip-lite');
        }

        if (this.cacheGeo[ip]) {
            return callback(null, this.cacheGeo[ip]);
        }

        callback(null, geoip.lookup(ip));
    }

    _geolocateFormatResult(data, result) {
        if (!this.options.geo) {
            return result;
        }

        result.city = '';
        result.country = '';
        result.region = '';
        result.latitude = '';
        result.longitude = '';

        if (!data) {
            return result;
        }

        this.cacheGeo[result.ip] = data;

        result.city = data.city || '';
        result.country = data.country || '';
        result.region = data.region || '';
        result.latitude = data.ll[0] || '';
        result.longitude = data.ll[1] || '';

        return result;
    }

    _reverseDns(ip, callback) {
        if (!this.options.reverse) {
            callback();
            return;
        }

        if (this.cacheDns[ip]) {
            callback(null, this.cacheDns[ip]);
            return;
        }

        dns.reverse(ip, (err, domains) => {
            if (err) {
                if (err.code == 'ENOTFOUND') {
                    callback(null);
                    return;
                }

                // unknow error
                this.emit('error', {
                    fnc:'lookupDns',
                    err
                });
                callback(null);
                return;
            }

            if (!domains.length) {
                callback(null);
                return;
            }

            callback(null, domains[0]);
        });
    }

    _reverseDnsFormatResult(data, result) {
        if (!this.options.reverse || !result) return result;
        result.reverse = '';
        if (!data) return result;
        result.reverse = data;
        this.cacheDns[result.ip] = data;
        return result;
    }

    _portScan(ip, port, callback) {
        if (!port) {
            return callback();
        }

        const t = new TcpScan({
            ip,
            port,
            banner : this.options.banner,
            bannerlen : this.options.bannerlen,
            timeout : this.options.timeout
        });

        t.analyzePort(callback);
    }

    _portScanFormatResult(data, result) {

        if (!result || !this.options.port || !data) {
            return result;
        }

        if (!this.options.showTimeout && data.status.match(/timeout/i)) {
            return result = null;
        }

        if (!this.options.showRefuse && data.status.match(/refuse/i)) {
            return result = null;
        }

        if (!this.options.showOpen && data.status.match(/open/i)) {
            return result = null;
        }

        if (!this.options.showUnreachable && data.status.match(/unreachable/i)) {
            return result = null;
        }

        if (this.options.banner) {
            result.banner = data.banner || '';
        }

        if (this.options.bannerraw) {
            result.bannerraw = data.raw || '';
        }

        result.status = data.status;

        return result;
    }

    _formatFinalResult(result) {

        if (!result) {
            return result;
        }

        if (this.options.reverse && this.options.reversevalid && result.reverse == '') {
            result = null;
        }

        if (!result.port) delete result.port;

        if (this.options.json) {

            if (result.status && result.status.match(/close/i)) {
                delete result.banner;
                delete result.bannerraw;
            }

            if (!result.reverse) {
                delete result.reverse;
            }

            if (!result.banner) {
                delete result.banner;
            }

            if (!result.bannerraw) {
                delete result.bannerraw;
            }
        }
        return result;
    }

    scan(args, nextIteration) {

        let result = {
            ip:args.ip,
            port:args.port
        };

        async.series([
            (next) => {
                this._geolocate(args.ip, next);
            },
            (next) => {
                this._reverseDns(args.ip, next);
            },
            (next) => {
                this._portScan(args.ip, args.port, next);
            }
        ], (err, arr) => {
            result = this._geolocateFormatResult(arr[0], result);
            result = this._reverseDnsFormatResult(arr[1], result);
            result = this._portScanFormatResult(arr[2], result);
            result = this._formatFinalResult(result);
            if (result) {
                this.emit('result', result);
            }
            nextIteration();
        });
    }

    run(callback) {
        this.callbackRun = callback;
        this.q && this.q.run();
        return;
    }

    getOptions() {
        return this.options;
    }

    pause() {
        if (this.paused) return;
        this.paused = true;
        this.q.pause(true);
    }

    unpause() {
        if (!this.paused) return;
        this.paused = false;
        this.q.pause(false);
    }

    abort() {
        this.q.abort();
        return;
    }

    getInfo() {
        return this.info;
    }

}

module.exports = Evilscan;
